build_daemon/lib/client.dart

```diff
@@ -0,0 +1,129 @@
+    var request = BuildOptionsRequest((b) => b..options.replace(options));
+    _channel.sink.add(jsonEncode(serializers.serialize(request)));
+  }
+
+  /// Adds paths to write usage logs to.
+  ///
+  /// When the client disconnects these files will no longer be logged to.
+  void logToPaths(Iterable<String> paths) {
+    var request = LogToPathsRequest((b) => b..paths.replace(paths));
+    _channel.sink.add(jsonEncode(serializers.serialize(request)));
+  }
+
+  /// Registers a build target to be built upon any file change.
+  ///
+  /// Changes that match the patterns in [blackListPattern] will be ignored.
+  ///
+  /// Note that you should provide the "fully qualified" target name. For
+  /// example one should provide //some-target:test_run instead of
+  /// //some-target:test.
+  void registerBuildTarget(String target, Iterable<String> blackListPattern) {
+    var request = BuildTargetRequest((b) => b
+      ..target = target
+      ..blackListPattern.replace(blackListPattern));
+    _channel.sink.add(jsonEncode(serializers.serialize(request)));
+  }
+
+  /// Builds all registered targets.
+  void startBuild() {
+    var request = BuildRequest();
+    _channel.sink.add(jsonEncode(serializers.serialize(request)));
+  }
+
+  Future<void> _connect(String workingDirectory, int port) async {
+    _channel = IOWebSocketChannel.connect('ws://localhost:$port');
+    _channel.stream.listen(_handleServerMessage)
+      // TODO(grouma) - Implement proper error handling.
+      ..onError(print);
+  }
+
+  void _handleServerMessage(dynamic data) {
+    var message = serializers.deserialize(jsonDecode(data as String));
+    if (message is ServerLog) {
+      _serverLogStreamController.add(message);
+    } else if (message is BuildResults) {
+      _buildResultsStreamController.add(message);
+    } else {
+      // In practice we should never reach this state due to the
+      // deserialize call.
+      throw StateError(
+          'Unexpected message from the Dart Build Daemon\n $message');
+    }
+  }
+
+  static Future<BuildDaemonClient> connect(String workingDirectory,
+      {String daemonCommand = ''}) async {
+    Process process;
+    if (daemonCommand.isEmpty) {
+      process = await Process.start(
+          'pub', ['run', 'build_daemon', workingDirectory],
+          mode: ProcessStartMode.detachedWithStdio);
+    } else {
+      process = await Process.start(daemonCommand, [workingDirectory],
+          mode: ProcessStartMode.detachedWithStdio);
+    }
+    // Print errors coming from the Dart Build Daemon to help with debugging.
+    process.stderr.transform(utf8.decoder).listen(print);
+    var result = await process.stdout
+        .transform(utf8.decoder)
+        .transform(const LineSplitter())
+        .firstWhere((line) => line == versionSkew || line == readyToConnectLog);
+
+    if (result == versionSkew) {
+      throw VersionSkew();
+    }
+
+    var port = _existingPort(workingDirectory);
+    if (port == null) return null;
```

Comment:
Based on the example usage, maybe this should throw an exception instead?

Reply:
Changed.